### ISSUE TITLE
DynamoDB: Handle container types correctly

### ISSUE THREAD

**Issue by amotl** [crate/commons-codec#33]
## About
Coming from a fix to handle map/object types correctly, and corresponding attempts to introduce CrateDB SQL casts to get it right, we may have spotted an anomaly when processing JSON data on OBJECTs vs. ARRAYs.

## Details
Do you think this is an anomaly that deserves a ticket on crate/crate? The first statement casts an object successfully from JSON, while the second one croaks on the list/array.

```sql
cr> select '{"foo": "bar"}'::OBJECT;
+----------------+
| {"foo"='bar'}  |
+----------------+
| {"foo": "bar"} |
+----------------+
```

```sql
select '["foobar"]'::ARRAY(TEXT);
SQLParseException[Cannot cast `'["foobar"]'` of type `text` to type `text_array`]
```

## References
- GH-32
- GH-31

**Comment by amotl** [comment]
@seut responded:

> We cannot cast a JSON->Array. Without quotes it should work because it is then parsed as an array literal instead of text.

@amotl said:

> Got it.
> ```
> cr> select ['foobar']::ARRAY(TEXT);
> +------------+
> | ['foobar'] |
> +------------+
> | ["foobar"] |
> +------------+
> ```

**Comment by amotl** [comment]
First of all, thanks for your elaborations.

This makes it complicated if an ARRAY contains OBJECTs, because then we cannot “simply” serialize them to JSON, but have to use the special syntax, for which there is of course no ready-made recursive serializer a la `json.dumps()`.

Is there any chance CrateDB could provide symmetric support on the SQL level to handle both container data types OBJECT and ARRAY equally well?

**Comment by seut** [comment]
I one should cast to the `JSON` type when working with json data instead of using our dedicated container structures. 
This already works with top-level arrays as well:

```
cr> select '["foobar"]'::json;
+--------------+
| '["foobar"]' |
+--------------+
| ["foobar"]   |
+--------------+
SELECT 1 row in set (0.012 sec)
```

**Comment by amotl** [comment]
> So the only solution seems to support the JSON -> ARRAY cast.

Thank you so much for your evaluations. So, until CrateDB will gain that feature, we need to use the non-JSON syntax to describe ARRAYs.

### Primitive inner types
It works well for those.
```sql
cr> UPDATE "foo" SET data['tags'] = ['foo', 'bar'];
UPDATE OK, 1 row affected (0.049 sec)
cr> select pg_typeof(data['tags']) from foo;
+------------------------------------+
| pg_catalog.pg_typeof(data['tags']) |
+------------------------------------+
| text_array                         |
+------------------------------------+
```

### Complex inner types
It also works well for those.
```sql
cr> UPDATE "foo" SET data['supertags'] = [{foo='bar', baz='qux'}];
UPDATE OK, 1 row affected (0.052 sec)
cr> select pg_typeof(data['supertags']) from foo;
+-----------------------------------------+
| pg_catalog.pg_typeof(data['supertags']) |
+-----------------------------------------+
| object_array                            |
+-----------------------------------------+
```

So, I figure it's a bummer, but not a blocker.

**Comment by amotl** [comment]
Fixed with GH-34, and released with v0.0.11.
